cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java

```diff
@@ -0,0 +1,107 @@
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.UserErrors;
+import co.cask.cdap.internal.UserMessages;
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.internal.app.services.PropertiesResolver;
+import co.cask.cdap.messaging.MessagingService;
+import co.cask.cdap.messaging.client.StoreRequestBuilder;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Task publisher that sends notification for a triggered schedule.
+ */
+public final class ScheduleTaskPublisher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScheduleTaskPublisher.class);
+
+  private static final Gson GSON = new Gson();
+
+  private final MessagingService messagingService;
+  private final TopicId topicId;
+  private final Store store;
+  private final PropertiesResolver propertiesResolver;
+  private final RunConstraintsChecker requirementsChecker;
+  public ScheduleTaskPublisher(Store store, MessagingService messagingService,
+                            PropertiesResolver propertiesResolver, CConfiguration cConf) {
+    this.store = store;
+    this.messagingService = messagingService;
+    this.propertiesResolver = propertiesResolver;
+    this.requirementsChecker = new RunConstraintsChecker(store);
+    this.topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(), cConf.get(Constants.Scheduler.TIME_EVENT_TOPIC));
+  }
+
+  /**
+   * Checks if all schedule requirements are satisfied,
+   * then publish notification to trigger the corresponding schedule.
+   *
+   * @param programId Program Id
+   * @param systemOverrides Arguments that would be supplied as system runtime arguments for the program.
+   * @param userOverrides Arguments to add to the user runtime arguments for the program.
+   */
+  public void publishNotification(ProgramId programId, Map<String, String> systemOverrides,
+                                 Map<String, String> userOverrides) throws Exception {
+    Map<String, String> userArgs = Maps.newHashMap();
+    Map<String, String> systemArgs = Maps.newHashMap();
+
+    String scheduleName = systemOverrides.get(ProgramOptionConstants.SCHEDULE_NAME);
+    ApplicationSpecification appSpec = store.getApplication(programId.getParent());
+    if (appSpec == null) {
+      throw new TaskExecutionException(String.format(UserMessages.getMessage(UserErrors.PROGRAM_NOT_FOUND), programId),
+                                       false);
+    }
+
+    ScheduleSpecification spec = appSpec.getSchedules().get(scheduleName);
+    if (!requirementsChecker.checkSatisfied(programId, spec.getSchedule())) {
+      return;
```

Comment:
why would you check the run constraints here? All you should do it publish. The scheduler's subscriber will receive the notification and create a new job in the job queue. Then it will check the constraints before  the job is launched. 

If you check here, then the job will not ever be launched, even if the constraints are satisfied 5 minutes later. Also, it is better to have the constraint checking in a single place, for separation of concerns